<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Statistical Framework for Modeling and Identifying Differential 
        Distributions in Single-cell RNA-sequencing Data â€¢ scShapes</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Statistical Framework for Modeling and Identifying Differential 
        Distributions in Single-cell RNA-sequencing Data">
<meta property="og:description" content="We present a novel statistical framework for identifying 
        differential distributions in single-cell RNA-sequencing 
        (scRNA-seq) data between treatment conditions by modeling gene 
        expression read counts using generalized linear models (GLMs). 
        We model each gene independently under each treatment condition 
        using error distributions Poisson, Negative Binomial (NB), 
        Zero-inflated Poisson (ZIP) and Zero-inflated Negative Binomial (ZINB) 
        with log link function and model based normalization for differences 
        in sequencing depth. Since all four distributions considered in 
        our framework belong to the same family of distributions, 
        we first perform a Kolmogorov-Smirnov (KS) test to select genes 
        belonging to the family of ZINB distributions. Genes passing the 
        KS test will be then modeled using GLMs. Model selection is done 
        by calculating the Bayesian Information Criterion (BIC) and likelihood 
        ratio test (LRT) statistic. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scShapes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/vignette_scShapes.html">The vignette for running  scShapes</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Malindrie/scShapes/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="scshapes" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#scshapes" class="anchor"></a>scShapes</h1></div>
<!-- badges: start -->

<p><a href="https://ci.appveyor.com/project/Malindrie/scShapes"><img src="https://ci.appveyor.com/api/projects/status/github/Malindrie/scShapes?branch=master&amp;svg=true" alt="AppVeyor build status"></a> <!-- badges: end --></p>
<p>We present a novel statistical framework for identifying differential distributions in single-cell RNA-sequencing (scRNA-seq) data between treatment conditions by modeling gene expression read counts using generalized linear models (GLMs). We model each gene independently under each treatment condition using the error distributions Poisson (P), Negative Binomial (NB), Zero-inflated Poisson (ZIP) and Zero-inflated Negative Binomial (ZINB) with log link function and model based normalization for differences in sequencing depth. Since all four distributions considered in our framework belong to the same family of distributions, we first perform a Kolmogorov-Smirnov (KS) test to select genes belonging to the family of ZINB distributions. Genes passing the KS test will be then modeled using GLMs. Model selection is done by calculating the Bayesian Information Criterion and likelihood ratio test statistic.</p>
<p>While most methods for differential gene expression analysis aim to detect a shift in the mean of expressed values, single cell data are driven by over-dispersion and dropouts requiring statistical distributions that can handle the excess zeros. By modeling gene expression distributions, our framework can identify subtle variations that do not involve the change in mean. It also has the flexibility to adjust for covariates and perform multiple comparisons while explicitly modeling the variability between samples.</p>
<p>You can view the preprint of our method in <a href="https://www.biorxiv.org/content/10.1101/2022.02.13.480299v1">bioRxiv</a></p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the released version of scShapes with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">'Malindrie/scShapes'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Malindrie/scShapes">scShapes</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>This is a basic example which shows how you can use scShapes for identifying differential distributions in single-cell RNA-seq data. For this example data we use the human immune cells (PBMC) dataset distributed through the <a href="https://github.com/satijalab/seurat-data">SeuratData</a> package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Malindrie/scShapes">scShapes</a></span><span class="op">)</span>

<span class="co">#Loading and preparing data for input </span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat">SeuratData</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">0xBEEF</span><span class="op">)</span>

<span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html">InstallData</a></span><span class="op">(</span><span class="st">"ifnb"</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html">LoadData</a></span><span class="op">(</span><span class="st">"ifnb"</span><span class="op">)</span>
<span class="va">ifnb.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SplitObject.html">SplitObject</a></span><span class="op">(</span><span class="va">ifnb</span>, split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></code></pre></div>
<p>We first filter the genes to keep only genes expressed in at least 10% of cells:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#First extract the RNA-seq counts from the 'RNA' assay of the seurat object</span>
<span class="va">ifnb.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">ifnb.list</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">x</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>
<span class="va">ifnb.filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">ifnb.obj</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="reference/filter_counts.html">filter_counts</a></span><span class="op">(</span><span class="va">x</span>, perc.zero <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Removing 527 rows of genes with all zero counts</span>
<span class="co">#&gt; Removing 778 rows of genes with all zero counts</span></code></pre></div>
<p>In order to normalize for differences in sequencing depth, the log of the total UMI counts assigned per cell will be used as an offset in the GLM. This function is inbuilt in the algorithm; however the user is required to input the library sizes. We can calculate the library sizes for the two treatment conditions as;</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ifnb.lib.size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">ifnb.filtered</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The â€˜meta.dataâ€™ slot of the Seurat object also contains information on the cell-types, which will be used as a covariate in the GLM model to account for known biological variation in the data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ifnb.variables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">ifnb.list</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
                        cell.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">x</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_annotations</span><span class="op">)</span>,
                        row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>For the purpose of this example we only run the pipeline for randomly selected 20 common genes under both treatment conditions â€˜CTRLâ€™ and â€˜STIMâ€™.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Randomly select 20 genes among common genes between the two treatment conditions</span>
<span class="va">comm.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">CTRL</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">STIM</span><span class="op">)</span><span class="op">)</span>
<span class="va">comm.20.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">comm.genes</span>, <span class="fl">20</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co">#Subset the randomly selected 20 genes</span>
<span class="va">ifnb.ctrl</span> <span class="op">&lt;-</span> <span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">CTRL</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">CTRL</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">comm.20.genes</span>,<span class="op">]</span>
<span class="va">ifnb.stim</span> <span class="op">&lt;-</span> <span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">STIM</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">STIM</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">comm.20.genes</span>,<span class="op">]</span>
<span class="va">ifnb.subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>CTRL <span class="op">=</span> <span class="va">ifnb.ctrl</span>, STIM <span class="op">=</span> <span class="va">ifnb.stim</span><span class="op">)</span></code></pre></div>
<p>Perform Kolmogorov-Smirnov test to select genes belonging to the family of ZINB distributions.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ifnb.ctrl.KS</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ks_test.html">ks_test</a></span><span class="op">(</span><span class="va">ifnb.subset</span><span class="op">$</span><span class="va">CTRL</span>, cexpr<span class="op">=</span><span class="va">ifnb.variables</span><span class="op">$</span><span class="va">CTRL</span>, lib.size<span class="op">=</span><span class="va">ifnb.lib.size</span><span class="op">$</span><span class="va">CTRL</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span>workers<span class="op">=</span><span class="fl">8</span>,type<span class="op">=</span><span class="st">"SOCK"</span><span class="op">)</span><span class="op">)</span>
<span class="va">ifnb.stim.KS</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ks_test.html">ks_test</a></span><span class="op">(</span><span class="va">ifnb.subset</span><span class="op">$</span><span class="va">STIM</span>, cexpr<span class="op">=</span><span class="va">ifnb.variables</span><span class="op">$</span><span class="va">STIM</span>, lib.size<span class="op">=</span><span class="va">ifnb.lib.size</span><span class="op">$</span><span class="va">STIM</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span>workers<span class="op">=</span><span class="fl">8</span>,type<span class="op">=</span><span class="st">"SOCK"</span><span class="op">)</span><span class="op">)</span>

<span class="co">#Select genes significant from the KS test.</span>
<span class="co">#By default the 'ks_sig' function performs Benjamini-Hochberg correction for multiple hypothese testing</span>
<span class="co">#and selects genes significant at p-value of 0.01</span>

<span class="va">ifnb.ctrl.sig.KS</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ks_sig.html">ks_sig</a></span><span class="op">(</span><span class="va">ifnb.ctrl.KS</span><span class="op">)</span>
<span class="va">ifnb.stim.sig.KS</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ks_sig.html">ks_sig</a></span><span class="op">(</span><span class="va">ifnb.stim.KS</span><span class="op">)</span>

<span class="co">#Subset UMI counts corresponding to the genes significant from the KS test</span>
<span class="va">ifnb.sig.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>CTRL <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">ifnb.ctrl.sig.KS</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span>,
                       STIM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">ifnb.stim.sig.KS</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span><span class="op">)</span>
<span class="va">ifnb.ctrl.KS</span> <span class="op">&lt;-</span> <span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">CTRL</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">CTRL</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ifnb.sig.genes</span><span class="op">$</span><span class="va">CTRL</span><span class="op">)</span>,<span class="op">]</span>
  <span class="va">ifnb.stim.KS</span> <span class="op">&lt;-</span> <span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">STIM</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ifnb.filtered</span><span class="op">$</span><span class="va">STIM</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ifnb.sig.genes</span><span class="op">$</span><span class="va">STIM</span><span class="op">)</span>,<span class="op">]</span></code></pre></div>
<p>Fit the 4 distributions P,NB,ZIP,ZINB for genes that belong to the ZINB family of distributions by fitting GLM with log of the library sizes as an offset and cell types as a covariate in the GLM.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ifnb.ctrl.fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/fit_models.html">fit_models</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">ifnb.ctrl.KS</span>, cexpr<span class="op">=</span><span class="va">ifnb.variables</span><span class="op">$</span><span class="va">CTRL</span>, lib.size<span class="op">=</span><span class="va">ifnb.lib.size</span><span class="op">$</span><span class="va">CTRL</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span>workers<span class="op">=</span><span class="fl">2</span>,type<span class="op">=</span><span class="st">"SOCK"</span><span class="op">)</span><span class="op">)</span>
<span class="va">ifnb.stim.fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/fit_models.html">fit_models</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">ifnb.stim.KS</span>, cexpr<span class="op">=</span><span class="va">ifnb.variables</span><span class="op">$</span><span class="va">STIM</span>, lib.size<span class="op">=</span><span class="va">ifnb.lib.size</span><span class="op">$</span><span class="va">STIM</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span>workers<span class="op">=</span><span class="fl">2</span>,type<span class="op">=</span><span class="st">"SOCK"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Once the 4 distributions are fitted, we next calculate the BIC value for each model and select the model with the least BIC value.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ifnb.ctrl.bic.val</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/model_bic.html">model_bic</a></span><span class="op">(</span><span class="va">ifnb.ctrl.fit</span><span class="op">)</span>
<span class="va">ifnb.stim.bic.val</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/model_bic.html">model_bic</a></span><span class="op">(</span><span class="va">ifnb.stim.fit</span><span class="op">)</span>

<span class="co">#select model with least bic value</span>
<span class="va">ifnb.ctrl.lbic</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/lbic_model.html">lbic_model</a></span><span class="op">(</span><span class="va">ifnb.ctrl.bic.val</span>, <span class="va">ifnb.ctrl.KS</span><span class="op">)</span>
<span class="va">ifnb.stim.lbic</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/lbic_model.html">lbic_model</a></span><span class="op">(</span><span class="va">ifnb.stim.bic.val</span>, <span class="va">ifnb.stim.KS</span><span class="op">)</span></code></pre></div>
<p>To ensure the fit of the models selected based on the least BIC value, additionally we perform LRT to test for model adequacy and presence of zero-inflation.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ifnb.ctrl.gof</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/gof_model.html">gof_model</a></span><span class="op">(</span><span class="va">ifnb.ctrl.lbic</span>, <span class="va">ifnb.variables</span><span class="op">$</span><span class="va">CTRL</span>, <span class="va">ifnb.lib.size</span><span class="op">$</span><span class="va">CTRL</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">ifnb.stim.gof</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/gof_model.html">gof_model</a></span><span class="op">(</span><span class="va">ifnb.stim.lbic</span>, <span class="va">ifnb.variables</span><span class="op">$</span><span class="va">STIM</span>, <span class="va">ifnb.lib.size</span><span class="op">$</span><span class="va">STIM</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Finally based on the results of the model adequacy tests, we can identify the distribution of best fit for each gene.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ifnb.ctrl.dist.fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/select_model.html">select_model</a></span><span class="op">(</span><span class="va">ifnb.ctrl.gof</span><span class="op">)</span>
<span class="va">ifnb.stim.dist.fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/select_model.html">select_model</a></span><span class="op">(</span><span class="va">ifnb.stim.gof</span><span class="op">)</span></code></pre></div>
<p>Once the distribution of best fit is identified for genes of interest, it is also possible to extract parameters of interest for the models.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ifnb.ctrl.params</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/model_param.html">model_param</a></span> <span class="op">(</span><span class="va">ifnb.ctrl.fit</span>, <span class="va">ifnb.ctrl.dist.fit</span>, model<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span>
<span class="va">ifnb.stim.params</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/model_param.html">model_param</a></span> <span class="op">(</span><span class="va">ifnb.stim.fit</span>, <span class="va">ifnb.stim.dist.fit</span>, model<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></code></pre></div>
<p>Using above results we can now identify the differentially distributed genes between â€˜CTRLâ€™ and â€˜STIMâ€™. First we need to subset genes that is significant in the KS test in both conditions.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Subset the common genes between the two groups, that pass the GOF test</span>
<span class="va">ifnb.ctrl.fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">ifnb.ctrl.dist.fit</span><span class="op">)</span>
<span class="va">ifnb.stim.fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">ifnb.stim.dist.fit</span><span class="op">)</span>
<span class="va">ifnb.gof.sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">intersect</a></span><span class="op">(</span><span class="va">ifnb.ctrl.fit</span>, <span class="va">ifnb.stim.fit</span><span class="op">)</span>

<span class="va">ifnb.dist.ctrl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">ifnb.ctrl.dist.fit</span><span class="op">$</span><span class="va">P_genes</span>, <span class="va">ifnb.ctrl.dist.fit</span><span class="op">$</span><span class="va">NB_genes</span>, <span class="va">ifnb.ctrl.dist.fit</span><span class="op">$</span><span class="va">ZIP_genes</span>, <span class="va">ifnb.ctrl.dist.fit</span><span class="op">$</span><span class="va">ZINB_genes</span><span class="op">)</span><span class="op">)</span>
<span class="va">ifnb.dist.ctrl</span><span class="op">$</span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Po"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ifnb.ctrl.dist.fit</span><span class="op">$</span><span class="va">P_genes</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"NB"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ifnb.ctrl.dist.fit</span><span class="op">$</span><span class="va">NB_genes</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"ZIP"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ifnb.ctrl.dist.fit</span><span class="op">$</span><span class="va">ZIP_genes</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"ZINB"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ifnb.ctrl.dist.fit</span><span class="op">$</span><span class="va">ZINB_genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">ifnb.dist.stim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">ifnb.stim.dist.fit</span><span class="op">$</span><span class="va">P_genes</span>, <span class="va">ifnb.stim.dist.fit</span><span class="op">$</span><span class="va">NB_genes</span>, <span class="va">ifnb.stim.dist.fit</span><span class="op">$</span><span class="va">ZIP_genes</span>, <span class="va">ifnb.stim.dist.fit</span><span class="op">$</span><span class="va">ZINB_genes</span><span class="op">)</span><span class="op">)</span>
<span class="va">ifnb.dist.stim</span><span class="op">$</span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Po"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ifnb.stim.dist.fit</span><span class="op">$</span><span class="va">P_genes</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"NB"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ifnb.stim.dist.fit</span><span class="op">$</span><span class="va">NB_genes</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"ZIP"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ifnb.stim.dist.fit</span><span class="op">$</span><span class="va">ZIP_genes</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"ZINB"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ifnb.stim.dist.fit</span><span class="op">$</span><span class="va">ZINB_genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co">#Dataframe consisting of distributions followed by each gene passing the KS test</span>
<span class="va">ifnb.gof.ctrl</span> <span class="op">&lt;-</span> <span class="va">ifnb.dist.ctrl</span><span class="op">[</span><span class="va">ifnb.dist.ctrl</span><span class="op">$</span><span class="va">gene</span> <span class="op">%in%</span> <span class="va">ifnb.gof.sig</span>,<span class="op">]</span>
<span class="va">ifnb.gof.stim</span> <span class="op">&lt;-</span> <span class="va">ifnb.dist.stim</span><span class="op">[</span><span class="va">ifnb.dist.stim</span><span class="op">$</span><span class="va">gene</span> <span class="op">%in%</span> <span class="va">ifnb.gof.sig</span>,<span class="op">]</span>

<span class="va">ifnb.distr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>ctrl <span class="op">=</span> <span class="va">ifnb.gof.ctrl</span><span class="op">$</span><span class="va">dist</span>, row.names <span class="op">=</span> <span class="va">ifnb.gof.ctrl</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>
<span class="va">ifnb.distr</span><span class="op">$</span><span class="va">stim</span> <span class="op">&lt;-</span> <span class="va">ifnb.gof.stim</span><span class="op">$</span><span class="va">dist</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ifnb.distr</span><span class="op">)</span>, <span class="va">ifnb.gof.stim</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>Using the dataframe of genes and distribution followed under each condition now we can identify genes changing distribution between â€˜CTRLâ€™ and â€˜STIMâ€™</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ifnb.DD.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/change_shape.html">change_shape</a></span><span class="op">(</span><span class="va">ifnb.distr</span><span class="op">)</span></code></pre></div>
<p>This will give a list of two lists with genes changing distribution between condition and genes changing distribution from unimodal in one condition to zero-inflated in the other condition.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/scShapes">https://â€‹www.bioconductor.org/â€‹packages/â€‹scShapes</a>
</li>
<li>Browse source code at <br><a href="https://github.com/Malindrie/scShapes/">https://â€‹github.com/â€‹Malindrie/â€‹scShapes/â€‹</a>
</li>
<li>Report a bug at <br><a href="https://github.com/Malindrie/scShapes/issues">https://â€‹github.com/â€‹Malindrie/â€‹scShapes/â€‹issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Malindrie Dharmaratne <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0002-1694-6496" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.com/Malindrie/scShapes"><img src="https://travis-ci.com/Malindrie/scShapes.svg?branch=master" alt="Travis build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Malindrie Dharmaratne.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
