<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>The vignette for running  scShapes â€¢ scShapes</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="The vignette for running  scShapes">
<meta property="og:description" content="scShapes">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scShapes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette_scShapes.html">The vignette for running  scShapes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Malindrie/scShapes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="vignette_scShapes_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>The vignette for running scShapes</h1>
                        <h4 data-toc-skip class="author">Malindrie Dharmaratne</h4>
            
            <h4 data-toc-skip class="date">2022-11-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Malindrie/scShapes/blob/HEAD/vignettes/vignette_scShapes.Rmd" class="external-link"><code>vignettes/vignette_scShapes.Rmd</code></a></small>
      <div class="hidden name"><code>vignette_scShapes.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="pulling-and-running-singularity-container">Pulling and running singularity container<a class="anchor" aria-label="anchor" href="#pulling-and-running-singularity-container"></a>
</h3>
<pre><code>$ singularity pull --name scShapes.sif shub://Malindrie/scShapes
$ singularity shell scShapes.sif
Singularity scShapes.sif:~ &gt; </code></pre>
</div>
<div class="section level3">
<h3 id="pulling-and-running-docker-container">Pulling and running docker container<a class="anchor" aria-label="anchor" href="#pulling-and-running-docker-container"></a>
</h3>
<pre><code>$ docker pull maldharm/scshapes
$ docker run -it maldharm/scshapes
root@516e456b0762:/#</code></pre>
<p>To use containerized <code>scShapes</code> package, need to launch <code>R</code> or execute an R script through <code>Rscript</code>.</p>
</div>
<div class="section level3">
<h3 id="install-scshapes-from-install_github">Install <code>scShapes</code> from <code>install_github</code><a class="anchor" aria-label="anchor" href="#install-scshapes-from-install_github"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">'Malindrie/scShapes'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Malindrie/scShapes" class="external-link">scShapes</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0xBEEF</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h3>
<p>This is a basic example which shows how you can use scShapes for identifying differential distributions in single-cell RNA-seq data. For this example data we use the toy example data <code>scData</code> included in the package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Loading and preparing data for input </span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">scData</span><span class="op">)</span></span></code></pre></div>
<p>We first filter the genes to keep only genes expressed in at least 10% of cells:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">scData_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_counts.html">filter_counts</a></span><span class="op">(</span><span class="va">scData</span><span class="op">$</span><span class="va">counts</span>, perc.zero <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p>In order to normalize for differences in sequencing depth, the log of the total UMI counts assigned per cell will be used as an offset in the GLM. This function is inbuilt in the algorithm; however the user is required to input the library sizes. In our example data this information together with the covariates is given under the lists <code>lib_size</code> and <code>covariates</code> respectively.</p>
<p>Perform Kolmogorov-Smirnov test to select genes belonging to the family of ZINB distributions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scData_KS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ks_test.html">ks_test</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">scData</span><span class="op">$</span><span class="va">counts</span>, cexpr<span class="op">=</span><span class="va">scData</span><span class="op">$</span><span class="va">covariates</span>, lib.size<span class="op">=</span><span class="va">scData</span><span class="op">$</span><span class="va">lib_size</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span>workers<span class="op">=</span><span class="fl">8</span>,type<span class="op">=</span><span class="st">"SOCK"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select genes significant from the KS test.</span></span>
<span><span class="co"># By default the 'ks_sig' function performs Benjamini-Hochberg correction for multiple   hypothese testing</span></span>
<span><span class="co"># and selects genes significant at p-value of 0.01</span></span>
<span></span>
<span><span class="va">scData_KS_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ks_sig.html">ks_sig</a></span><span class="op">(</span><span class="va">scData_KS</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset UMI counts corresponding to the genes significant from the KS test</span></span>
<span><span class="va">scData.sig.genes</span> <span class="op">&lt;-</span> <span class="va">scData</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scData</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scData_KS_sig</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<p>Fit the 4 distributions P,NB,ZIP,ZINB for genes that belong to the ZINB family of distributions by fitting GLM with log of the library sizes as an offset and cell types as a covariate in the GLM.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scData_models</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_models.html">fit_models</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">scData.sig.genes</span>, cexpr<span class="op">=</span><span class="va">scData</span><span class="op">$</span><span class="va">covariates</span>, lib.size<span class="op">=</span><span class="va">scData</span><span class="op">$</span><span class="va">lib_size</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span>workers<span class="op">=</span><span class="fl">8</span>,type<span class="op">=</span><span class="st">"SOCK"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Once the 4 distributions are fitted, we next calculate the BIC value for each model and select the model with the least BIC value.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scData_bicvals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/model_bic.html">model_bic</a></span><span class="op">(</span><span class="va">scData_models</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select model with least bic value</span></span>
<span><span class="va">scData_least.bic</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lbic_model.html">lbic_model</a></span><span class="op">(</span><span class="va">scData_bicvals</span>, <span class="va">scData</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span></code></pre></div>
<p>To ensure the fit of the models selected based on the least BIC value, additionally we perform LRT to test for model adequacy and presence of zero-inflation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scData_gof</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gof_model.html">gof_model</a></span><span class="op">(</span><span class="va">scData_least.bic</span>, cexpr<span class="op">=</span><span class="va">scData</span><span class="op">$</span><span class="va">covariates</span>, lib.size<span class="op">=</span><span class="va">scData</span><span class="op">$</span><span class="va">lib_size</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span>workers<span class="op">=</span><span class="fl">8</span>,type<span class="op">=</span><span class="st">"SOCK"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Finally based on the results of the model adequacy tests, we can identify the distribution of best fit for each gene.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scData_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_model.html">select_model</a></span><span class="op">(</span><span class="va">scData_gof</span><span class="op">)</span></span></code></pre></div>
<p>Once the distribution of best fit is identified for genes of interest, it is also possible to extract parameters of interest for the models.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scData_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/model_param.html">model_param</a></span><span class="op">(</span><span class="va">scData_models</span>, <span class="va">scData_fit</span>, model<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p>If our dataset consists of multiple conditions we can follow the above approach to identify the best fir distribution shape for each gene under each treatment condition (selecting the subset of genes common between conditions). Then using the dataframe of genes and distribution followed under each condition, now we can identify genes changing distribution between conditions.</p>
<p>For example suppose we follow above pipeline for scRNA-seq data on two treatment conditions â€˜CTRLâ€™ and â€˜STIMâ€™ and have identified the best distribution fit for each gene under each condition independently. Suppose the dataframe <code>ifnb.distr</code>; with genes as rows and columns as â€˜CTRLâ€™ and â€˜STIMâ€™ with the corresponding distribution name a particular gene follows, then we can identify genes changing distribution shape between â€˜CTRLâ€™ and â€˜STIMâ€™ as;</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb.DD.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/change_shape.html">change_shape</a></span><span class="op">(</span><span class="va">ifnb.distr</span><span class="op">)</span></span></code></pre></div>
<p>This will give a list of two lists with genes changing distribution between condition and genes changing distribution from unimodal in one condition to zero-inflated in the other condition.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>Here is the output of sessionInfo() on the system on which this document was compiled:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] BiocParallel_1.32.0 scShapes_1.0.0     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] Rcpp_1.0.9          plyr_1.8.7          bslib_0.4.1        </span></span>
<span><span class="co">#&gt;  [4] compiler_4.2.2      jquerylib_0.1.4     tools_4.2.2        </span></span>
<span><span class="co">#&gt;  [7] digest_0.6.30       jsonlite_1.8.3      evaluate_0.18      </span></span>
<span><span class="co">#&gt; [10] memoise_2.0.1       lattice_0.20-45     rlang_1.0.6        </span></span>
<span><span class="co">#&gt; [13] Matrix_1.5-1        cli_3.4.1           yaml_2.3.6         </span></span>
<span><span class="co">#&gt; [16] parallel_4.2.2      mvtnorm_1.1-3       pkgdown_2.0.6      </span></span>
<span><span class="co">#&gt; [19] xfun_0.34           fastmap_1.1.0       coda_0.19-4        </span></span>
<span><span class="co">#&gt; [22] stringr_1.4.1       knitr_1.40          desc_1.4.2         </span></span>
<span><span class="co">#&gt; [25] fs_1.5.2            sass_0.4.2          systemfonts_1.0.4  </span></span>
<span><span class="co">#&gt; [28] stats4_4.2.2        rprojroot_2.0.3     grid_4.2.2         </span></span>
<span><span class="co">#&gt; [31] R6_2.5.1            snow_0.4-4          textshaping_0.3.6  </span></span>
<span><span class="co">#&gt; [34] bdsmatrix_1.3-6     emdbook_1.3.12      VGAM_1.1-7         </span></span>
<span><span class="co">#&gt; [37] rmarkdown_2.17      purrr_0.3.5         magrittr_2.0.3     </span></span>
<span><span class="co">#&gt; [40] codetools_0.2-18    htmltools_0.5.3     MASS_7.3-58.1      </span></span>
<span><span class="co">#&gt; [43] splines_4.2.2       bbmle_1.0.25        numDeriv_2016.8-1.1</span></span>
<span><span class="co">#&gt; [46] dgof_1.4            ragg_1.2.4          stringi_1.7.8      </span></span>
<span><span class="co">#&gt; [49] pscl_1.5.5          cachem_1.0.6</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Malindrie Dharmaratne.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
